File: Ahorn/entities/BubbleDoor.jl
module XaphanHelperBubbleDoor

using ..Ahorn, Maple

@mapdef Entity "XaphanHelper/BubbleDoor" BubbleDoor(x::Integer, y::Integer, width::Integer=8, height::Integer=40, side::String="Left", directory::String="objects/XaphanHelper/BubbleDoor", color::String="Blue", flags::String="", forceLockedFlag::String="", openSound::String="", closeSound::String="",unlockSound::String="", lockSound::String="")

const placements = Ahorn.PlacementDict(
    "Bubble Door (Xaphan Helper)" => Ahorn.EntityPlacement(
        BubbleDoor,
        "point",
        ),
    )

    function Ahorn.selection(entity::BubbleDoor)
        x, y = Ahorn.position(entity)
        side = get(entity.data, "side", "Left")
        if side == "Left"
            return Ahorn.Rectangle(x, y, 8, 40)
        end
        if side == "Right"
            return Ahorn.Rectangle(x, y, 8, 40)
        end
        if side == "Top"
            return Ahorn.Rectangle(x, y, 40, 8)
        end
        if side == "Bottom"
            return Ahorn.Rectangle(x, y, 40, 8)
        end
    end

Ahorn.editingOptions(entity::BubbleDoor) = Dict{String, Any}(
    "side" => String["Left", "Right", "Top", "Bottom"],
    "color" => String["Blue", "Red", "Green", "Yellow", "Grey"]
)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::BubbleDoor, room::Maple.Room)
    side = get(entity.data, "side", "Left")
    color = get(entity.data, "color", "Blue")
    directory= get(entity.data, "directory", "objects/XaphanHelper/BubbleDoor")
    if side == "Left"
        Ahorn.drawSprite(ctx, "$(directory)/$(color)/closed00.png", 3, 20)
        Ahorn.drawSprite(ctx, "$(directory)/struct00.png", -4, 20)
    end
    if side == "Right"
        Ahorn.drawSprite(ctx, "$(directory)/$(color)/closed00.png", 11, 60, rot=pi)
        Ahorn.drawSprite(ctx, "$(directory)/struct00.png", 20, 60, rot=pi)
    end
    if side == "Top" 
        Ahorn.drawSprite(ctx, "$(directory)/$(color)/closed00.png", 43, 20, rot=pi/2)
        Ahorn.drawSprite(ctx, "$(directory)/struct00.png", 44, 12, rot=pi/2)
    end
    if side == "Bottom"
        Ahorn.drawSprite(ctx, "$(directory)/$(color)/closed00.png", 3, 28, rot=-pi/2)
        Ahorn.drawSprite(ctx, "$(directory)/struct00.png", 4, 36, rot=-pi/2)
    end
end

end

File: Loenn/entities/BubbleDoor.lua
local drawableSprite = require("structs.drawable_sprite")

local BubbleDoor = {}

BubbleDoor.name = "XaphanHelper/BubbleDoor"
BubbleDoor.depth = -9000
BubbleDoor.fieldOrder = {
    "x", "y", "directory", "side", "color", "flags", "forceLockedFlag", "openSound", "closeSound", "lockSound", "unlockSound", "onlyNeedOneFlag"
}
BubbleDoor.canResize = {false, false}
BubbleDoor.fieldInformation = {
    side = {
        options = {"Left", "Right", "Top", "Bottom"},
        editable = false
    },
    color = {
        options = {"Blue", "Red", "Green", "Yellow", "Grey"},
        editable = false
    }
}
BubbleDoor.placements = {
    name = "BubbleDoor",
    data = {
        side = "Left",
        directory = "objects/XaphanHelper/BubbleDoor",
        color = "Blue",
        flags = "",
        forceLockedFlag = "",
        openSound = "",
        closeSound = "",
        unlockSound = "",
        lockSound = "",
        onlyNeedOneFlag = false
    }
}

function BubbleDoor.sprite(room, entity)
    local side = entity.side or "Left"
    local color = entity.color or "Blue"
    local directory = entity.directory or "objects/XaphanHelper/BubbleDoor"
    
    local sprites = {}

    local sprite = drawableSprite.fromTexture(directory .. "/" .. color .. "/closed00", entity)
    local spriteStruct = drawableSprite.fromTexture(directory .. "/struct00", entity)

    if side == "Top" then
        sprite.rotation = math.pi / 2
        spriteStruct.rotation = math.pi / 2
        sprite:addPosition(20, 3)
        spriteStruct:addPosition(20, -4)
        sprite.scaleY = -1
        spriteStruct.scaleY = -1
    elseif side == "Bottom" then
        sprite.rotation = -math.pi / 2
        spriteStruct.rotation = -math.pi / 2
        sprite:addPosition(20, 5)
        spriteStruct:addPosition(20, 12)
    elseif side == "Left" then
        sprite.rotation = 0
        spriteStruct.rotation = 0
        sprite:addPosition(3, 20)
        spriteStruct:addPosition(-4, 20)
    elseif side == "Right" then
        sprite:addPosition(5, 20)
        spriteStruct:addPosition(12, 20)
        sprite.scaleX = -1
        spriteStruct.scaleX = -1
    end

    sprite:setColor("White")
    spriteStruct:setColor("White")

    if sprite then
        table.insert(sprites, sprite)
    end
    if spriteStruct then
        table.insert(sprites, spriteStruct)
    end

    return sprites
end

return BubbleDoor